await.go

```diff
@@ -100,101 +91,36 @@ func (a *PublicationAwaiter) Await(ctx context.Context, future IndexFuture) (Ind
 // the latest checkpoint from the log, parses the tree size, and releases all clients
 // that were blocked on an index smaller than this tree size.
 func (a *PublicationAwaiter) pollLoop(ctx context.Context, readCheckpoint func(ctx context.Context) ([]byte, error), pollPeriod time.Duration) {
-	for {
+	var (
+		cp     []byte
+		cpErr  error
+		cpSize uint64
+	)
+	for done := false; !done; {
 		select {
 		case <-ctx.Done():
 			klog.Info("PublicationAwaiter exiting due to context completion")
-			return
+			cp, cpSize, cpErr = nil, 0, ctx.Err()
+			done = true
 		case <-time.After(pollPeriod):
-			// It's worth this small lock contention to make sure that no unnecessary
-			// work happens in personalities that aren't performing writes.
-			a.mu.Lock()
-			hasClients := a.waiters.Front() != nil
-			a.mu.Unlock()
-			if !hasClients {
-				continue
+			cp, cpErr = readCheckpoint(ctx)
+			switch {
+			case errors.Is(cpErr, os.ErrNotExist):
+				return
+			case cpErr != nil:
+				cpSize = 0
+			default:
+				_, cpSize, _, cpErr = parse.CheckpointUnsafe(cp)
 			}
 		}
+
+		a.c.L.Lock()
 		// Note that for now, this releases all clients in the event of a single failure.
 		// If this causes problems, this could be changed to attempt retries.
-		rawCp, err := readCheckpoint(ctx)
-		if err != nil && !errors.Is(err, os.ErrNotExist) {
-			a.releaseClientsErr(fmt.Errorf("readCheckpoint: %v", err))
-			continue
-		}
-		_, size, _, err := parse.CheckpointUnsafe(rawCp)
-		if err != nil {
-			a.releaseClientsErr(err)
-			continue
-		}
-		a.releaseClients(size, rawCp)
+		a.checkpoint = cp
+		a.size = cpSize
+		a.err = cpErr
+		a.c.Broadcast()
```

Comment:
This broadcast seems likely to lead to a large CPU spike in high throughput situations, but worth trying it.